Implement DocValueAggregator for HLL DISTINCT.

BaurzhanSakhariev opened the pull request:
## Summary of the changes / Why this improves CrateDB

Result of running `./compare_run.py --v1 branch:master --v2 branch:baur/hll-docvalueaggregator --spec specs/select/hyperloglog.toml --forks 2 --env CRATE_HEAP_SIZE=2g -s path.data=/Users/baur/benchmark-data`

```
# Results (server side duration in ms)
V1: 4.6.0-41b3579b7049130cedd542077eabddb666ff7311
V2: 4.6.0-a40419673ac591c22d72252cea1d769b611699c3

Q: select hyperloglog_distinct("cCode") from uservisits
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |      422.632 ±  171.659 |    373.643 |    393.606 |    401.293 |   1596.161 |
|   V2    |      338.015 ±  166.400 |    300.876 |    304.461 |    314.744 |   1464.700 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -  22.25%                           -  25.54%   
There is a 98.60% probability that the observed difference is not random, and the best estimate of that difference is 22.25%
The test has statistical significance

Q: select hyperloglog_distinct("visitDate") from uservisits
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |      101.696 ±   36.188 |     90.875 |     96.032 |     98.174 |    351.191 |
|   V2    |       42.695 ±   22.822 |     37.574 |     38.818 |     39.590 |    199.010 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -  81.72%                           -  84.85%   
There is a 100.00% probability that the observed difference is not random, and the best estimate of that difference is 81.72%
The test has statistical significance


System/JVM Metrics (durations in ms, byte-values in MB)
    |    YOUNG GC            |       OLD GC           |      HEAP         |     ALLOC     
    |  cnt      avg      max |  cnt      avg      max |  initial     used |     rate      total
 V1 |   36     2.76     1.74 |   12    39.74    37.11 |     2147       66 |     0.00          0
 V2 |    7     3.56     2.70 |    3    37.83    40.72 |     2147      101 |     0.00          0
    
V1 top allocation frames
  ByteBuffer.get(...):2958
  MurmurHash3.hash64(...):567
  UnicodeUtil.UTF8toUTF16(...):519
  AggregateCollector.iterate(...):507
  InputRow.get(int):440
V2 top allocation frames
  ByteBuffer.get(...):2734
  MurmurHash3.hash64(...):441
  Buffer.position(int):320
  ScopedMemoryAccess.getByteInternal(...):316
  DocValuesAggregates.getRow(...):280
```

Updated results - hash method with offset, no array copy + separate implementation for integral types
```
V1: 4.6.0-5f40784cdcde85ce6a3cb5f7bade37911d1b5ec4
V2: 4.6.0-a5892e92235534dfff983c742b6b7d2c78602dd4

Q: select hyperloglog_distinct("cCode") from uservisits
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |      445.569 ±  141.618 |    405.673 |    421.316 |    433.674 |   1416.460 |
|   V2    |      295.680 ±  138.544 |    267.682 |    273.738 |    278.858 |   1254.050 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -  40.44%                           -  42.47%   
There is a 100.00% probability that the observed difference is not random, and the best estimate of that difference is 40.44%
The test has statistical significance

Q: select hyperloglog_distinct("visitDate") from uservisits
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |      128.962 ±   45.780 |    111.293 |    115.108 |    127.359 |    409.681 |
|   V2    |       50.332 ±   67.537 |     38.114 |     39.676 |     41.148 |    517.950 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -  87.71%                           -  97.47%   
There is a 100.00% probability that the observed difference is not random, and the best estimate of that difference is 87.71%
The test has statistical significance


System/JVM Metrics (durations in ms, byte-values in MB)
    |    YOUNG GC            |       OLD GC           |      HEAP         |     ALLOC     
    |  cnt      avg      max |  cnt      avg      max |  initial     used |     rate      total
 V1 |   36     2.67     3.36 |   11    34.78    40.44 |     2147      142 |     0.00          0
 V2 |    0     0.00     0.00 |    0     0.00     0.00 |     2147        0 |     0.00          0
    
V1 top allocation frames
  ByteBuffer.get(...):2510
  DataInput.readVInt():1027
  MurmurHash3.hash64(...):577
  AggregateCollector.iterate(...):564
  InputRow.get(int):533
V2 top allocation frames
  ByteBuffer.get(...):2682
  MurmurHash3.hash64(...):403
  DocValuesAggregates.getRow(...):395
  Lucene80DocValuesProducer$TermsDict.seekExact(long):318
  BufferedIndexInput.readByte(long):217

```

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

mfussenegger reviewed (commented):
I left some high level comments regarding the interface changes. 

Overall I think it is on the right track :+1:

BaurzhanSakhariev commented on extensions/functions/src/main/java/io/crate/operation/aggregation/HyperLogLogDistinctAggregation.java:
```diff
@@ -318,6 +378,10 @@ static Murmur3Hash getForType(DataType<?> dataType, boolean allOn4_1) {
 
         abstract long hash(Object val);
 
+        long hashByteArray(byte[] utf8Bytes) {
+            throw new UnsupportedOperationException("hash() method must be used for " + this.getClass().getCanonicalName());
+        }
+
```
hash is an argument for encapsulated in HllState `hyperLogLogPlusPlus.collect(hash)`, I can compute in DocValueAggregator but I have to expose some new method like 
```
var utf8Bytes = values.nextValue().bytes;
var hash = state.isAllOn4_1() 
          ?
      MurmurHash3.hash64(utf8Bytes, utf8Bytes.length)
           : 
      MurmurHash3.hash128(utf8Bytes, 0, utf8Bytes.length, 0, hash);  
state.addHash(hash); // new method
```
Problem here is that hash128 requires new Hash128 object and  currently it's created only once but if I  do it in DocValueAggregator I have to create it every time like hash128(., , new MurmurHash3.128());

To address that I will still expose new state.addhash but will update hash128 not to receive an instant which apparently just holder of 2 longs and wherever we use it we just use hash128.h1 - so I will just return h1 like hash64 does.

BaurzhanSakhariev replied:
Done. The only my concern hash128 case - as I'm calling hash directly, I'm creating 
new MurmurHash3.Hash128() every time as hash128 accepts that objects as a last argument. I could remove it completely and return long as we use hash.h1 anyway (or keep final instance in docvalueaggregator like Murmur3Hash.Bytes does but I didn't like neither of those options.

mfussenegger replied:
It would also be possible to have the DocValueAggregator implementation for Bytes hold onto a `Hash128` instance and re-use it.

BaurzhanSakhariev replied:
>DocValueAggregator implementation for Bytes hold
yep, that's the second option from above but I was not sure it's good idea to add some specific hash implementation in DVA. Then I found out that we just use static instance for Bytes64 and others and only Bytes is not the same and just made it similar to others.

BaurzhanSakhariev commented:
set default precision to 18 (maximal possible), error is still big.
```
cr> select count(name) from test;                                                                                                                                                                                       
+-------------+
| count(name) |
+-------------+
|      100000 |
+-------------+
SELECT 1 row in set (0.052 sec)
cr> select count(distinct name) from test;                                                                                                                                                                              
+-------------------------------------+
| collection_count(collect_set(name)) |a
+-------------------------------------+
|                               71097 |
+-------------------------------------+
SELECT 1 row in set (0.093 sec)a
cr> select hyperloglog_distinct(name) from test;                                                                                                                                                                        
+----------------------------+
| hyperloglog_distinct(name) |
+----------------------------+
|                      99874 |
+----------------------------+

```

BaurzhanSakhariev commented on server/src/main/java/io/crate/execution/engine/aggregation/DocValueAggregator.java:
```diff
@@ -26,11 +26,13 @@
 public interface DocValueAggregator<T> {
 
-    public T initialState(RamAccounting ramAccounting);
+    public T initialState(CollectTask collectTask, Literal... optionalParams);
```
Moved literals to getDocValueAggregator

mfussenegger commented on docs/appendices/release-notes/unreleased.rst:
```diff
@@ -94,6 +94,8 @@ Changes
 - Added support for directory-level wild card expansion for URIs passed to
   ``COPY FROM`` statements.
 
+- Performance improvement for hyperloglog_distinct function.
```
```suggestion
- Improved the performance of the :ref`hyperloglog_distinct
  <aggregation-hyperloglog-distinct>` aggregation function.
```

mfussenegger commented on extensions/functions/src/main/java/io/crate/operation/aggregation/HyperLogLogDistinctAggregation.java:
```diff
@@ -111,13 +119,12 @@ public HllState iterate(RamAccounting ramAccounting,
-            }
-            state.init(memoryManager, precision);
+
+        int precision = HyperLogLogPlusPlus.DEFAULT_PRECISION;
+        if (args.length > 1) {
+            precision = DataTypes.INTEGER.sanitizeValue(args[1].value());
         }
+        state = initIfNeeded(state, memoryManager, precision);
```
This will trigger an additional `args[1].value()` call even if the state is already initionalized. If `args[1]` is a function or column that could be expensive. 

What is the benefit of this change?

BaurzhanSakhariev replied:
Wanted to have one common method for `init` and use it in 3 places. I see, semantics are different, in case of DocValueAggregators it's done in `initialState` once and for regular aggregation we can handle args only in `iterate`. Reverted initialization for regular aggregation.

mfussenegger commented on extensions/functions/src/main/java/io/crate/operation/aggregation/HyperLogLogDistinctAggregation.java:
```diff
@@ -144,6 +151,71 @@ public Long terminatePartial(RamAccounting ramAccounting, HllState state) {
+                        // ref.bytes is used below since hash methods has a 'length' argument to control real length of the array
+                        // and ref.length is used to ignore unused bytes.
+                        var hash = state.isAllOn4_1() ?
+                            MurmurHash3.hash64(ref.bytes, ref.length)
+                            : MurmurHash3.hash128(ref.bytes, 0, ref.length, 0, Murmur3Hash.Bytes.INSTANCE.hash).h1;
```
The `offset` of the `BytesRef` could also be `> 0` in which case the logic here would probably be wrong.

mfussenegger replied:
Using `Murmur3Hash.Bytes.INSTANCE.hash` here will not be thread-safe. The hash128 has to be a local instance that can be created within this `BinaryDocValueAggregator` instance.

BaurzhanSakhariev replied:
Good catch, thanks!

mfussenegger commented on extensions/functions/src/main/java/io/crate/operation/aggregation/HyperLogLogDistinctAggregation.java:
```diff
@@ -352,6 +451,8 @@ long hash(Object val) {
 
         static class Bytes extends Murmur3Hash {
 
+            static final Bytes INSTANCE = new Bytes();
```
Please revert - this won't be thread-safe because the `MurmurHash3.Hash128` is mutated. 

My earlier comment was probably not clear. I meant that `MurmurHash3.Hash128 hash = new MurmurHash3.Hash128();` can be done within the specialized `DocValueAggregator` instance - not that we can re-use a `Bytes` instance.

mfussenegger commented on extensions/functions/src/test/java/io/crate/operation/aggregation/HyperLogLogDistinctAggregationIntegrationTest.java:
```diff
@@ -0,0 +1,51 @@
+    @Test
+    public void test_hyperloglog_distinct_string() {
+        execute("create table tbl (x text)");
+        int bulkSize = 10000;
+        Object[][] bulkArgs = new Object[bulkSize][];
+        for (int i = 0; i < bulkSize; i++) {
+            bulkArgs[i] = new Object[]{Integer.toString(i)};
+            // For simplicity of actual distinct computation all elements are distinct.
+            // If data generation gets changed than error computation below has to be updated.
+            // Currently actual number of distinct is equal to bulkSize.
+        }
```
What's the reason for the large bulk size? This test will be rather expensive.

Aren't the unit-tests enough to cover the new optimizations?

This seems to be partially a new functional/behavior test. Not sure if that is needed as long as we assert that optimized execution yields the same results as the previous executions.

mfussenegger replied:
Did you investigate why it is `null` there? We should ensure that the aggregation test framework tests the optimized execution paths.

BaurzhanSakhariev replied:
It seems like inserted String values are not flushed. I tried to throw an Exception on `partialResultWithDocValues == null && implementation != null` and found out that [Count(String)](https://github.com/crate/crate/blob/b8f5bfc1ae1d34b0426f2a01217fde663505225b/server/src/test/java/io/crate/execution/engine/aggregation/impl/CountAggregationTest.java#L104) and [Arbitrary(String)](https://github.com/crate/crate/blob/b8f5bfc1ae1d34b0426f2a01217fde663505225b/server/src/test/java/io/crate/execution/engine/aggregation/impl/ArbitraryAggregationTest.java#L110) are also failing with that check. DocValues [are empty](https://github.com/crate/crate/blob/b8f5bfc1ae1d34b0426f2a01217fde663505225b/server/src/main/java/io/crate/execution/engine/collect/DocValuesAggregates.java#L177) but only for `String` type.

mfussenegger replied:
The problem isn't that the records are not inserted, but that it doesn't use the optimized path because there are no doc-values available. 

The reason for that is that we create the mapping definition as `text`:

https://github.com/crate/crate/blob/a40419673ac591c22d72252cea1d769b611699c3/server/src/testFixtures/java/io/crate/operation/aggregation/AggregationTestCase.java#L393-L394

This should be `keyword` for strings that are not analyzed.

BaurzhanSakhariev replied:
it worked, thanks!

mfussenegger commented on server/src/main/java/io/crate/execution/engine/aggregation/impl/templates/SortedNumericDocValueAggregator.java:
```diff
@@ -18,39 +18,61 @@
  * software solely pursuant to the terms of the relevant commercial agreement.
  */
+/*
+ * Licensed to Crate.io GmbH ("Crate") under one or more contributor
```
Duplicate copyright header?

mfussenegger commented on server/src/main/java/io/crate/execution/engine/collect/DocValuesAggregates.java:
```diff
@@ -162,14 +167,19 @@
             }
 
             var aggregationReferences = new ArrayList<Symbol>(aggregation.inputs().size());
+            var literals = new ArrayList<Symbol>();
```
```suggestion
            var literals = new ArrayList<Literal<?>>();
```

mfussenegger commented on server/src/main/java/io/crate/execution/engine/collect/DocValuesAggregates.java:
```diff
@@ -162,14 +167,19 @@
-                    // We can extend this to instead return an adapter
-                    // to the normal aggregation implementation
-                    return null;
+                if (input.symbolType() == SymbolType.LITERAL) {
+                    literals.add(input);
+                } else {
```
```suggestion
                if (input instanceof Literal<?> literal) {
                    literals.add(literal);
                } else {
```

mfussenegger commented on server/src/main/java/io/crate/execution/engine/collect/DocValuesAggregates.java:
```diff
@@ -187,7 +197,8 @@
                 Symbols.typeView(aggregationReferences),
-                fieldTypes
+                fieldTypes,
+                literals.toArray(Literal[]::new)
```
Could probably pass in the list directly without conversion. This is probably the only production call-site, so we don't need varargs call-convenience and using the list probably avoids some unchecked warnings

BaurzhanSakhariev commented:
I added benchmark results to the PR description. However, compare_py produced 2 results, I picked up the last one. Judging by logs it did the whole cycle (switching branch, building, running 2 different queries) twice - probably because of  --forks 2.

DocValueAggregation looks better,  UnicodeUtil.UTF8toUTF16 allocation frame has gone

BaurzhanSakhariev commented:
Added DocValueAggregatr for integral types, added dedicated test case which was failing before when I had only 1 aggregator for internal + floating point numbers. Added test case with random type and values

mfussenegger commented on extensions/functions/src/main/java/io/crate/operation/aggregation/HyperLogLogDistinctAggregation.java:
```diff
@@ -144,6 +155,113 @@ public Long terminatePartial(RamAccounting ramAccounting, HllState state) {
+                // BytesRef has a property 'length' providing length of used bytes.
+                // Offset has to be taken to account as well.
+
+                byte[] bytes = new byte[ref.length];
+                System.arraycopy(ref.bytes, ref.offset, bytes, 0, ref.length);
+
+                var hash = state.isAllOn4_1() ?
+                    MurmurHash3.hash64(bytes, bytes.length)
+                    : MurmurHash3.hash128(bytes, 0, bytes.length, 0, hash128).h1;
```
How difficult would it be to adapt the hash functions to support a offset parameter to avoid the array copy?

BaurzhanSakhariev replied:
Updated hash64 method. Hash128 already works with offset

BaurzhanSakhariev replied:
Re-run becnhmark, updated PR description with the new result -  it got better, thanks.

Still don't understand why second conpary_py makes 2 rounds and second round is faster -  is it something like warmed up/cached version? Anyway, I picked up second round results in both cases (prevent and today's version)

mfussenegger replied:
It is due to the `--fork 2` and is to show that there can be fluctuations between runs.

BaurzhanSakhariev commented on server/src/testFixtures/java/io/crate/operation/aggregation/AggregationTestCase.java:
```diff
@@ -382,6 +398,9 @@ private XContentBuilder buildMapping(List<DataType<?>> argumentTypes) throws IOE
                         DataTypes.esMappingNameFrom(
                             ((ArrayType<?>) type).innerType().id()))
                     .endObject();
+            } else if(type.id() == StringType.ID) {
```
thx. `checkstyleMain` doesn't cover test files, does it?

mfussenegger commented on extensions/functions/src/main/java/io/crate/operation/aggregation/HyperLogLogDistinctAggregation.java:
```diff
@@ -144,6 +155,110 @@ public Long terminatePartial(RamAccounting ramAccounting, HllState state) {
+        public void apply(RamAccounting ramAccounting, int doc, HllState state) throws IOException {
+            if (values.advanceExact(doc) && values.docValueCount() == 1) {
+                BytesRef ref = values.nextValue();
+                // Using ref.bytes for getting utf-8 bytearray is wrong,
```
I think the comment could now be removed

mfussenegger commented on server/src/main/java/org/elasticsearch/common/hash/MurmurHash3.java:
```diff
@@ -65,34 +65,35 @@ protected static long fmix(long k) {
      */
-    public static long hash64(byte[] data, int length) {
-        return hash64(data, length, DEFAULT_SEED);
+    public static long hash64(byte[] data, int length, int offset) {
```
Maybe swap the arguments.  Most methods that work with bytes and support offset/length  have the offset parameter before the length parameter

BaurzhanSakhariev replied:
Swapped here and in further method with seed as well.

mfussenegger reviewed (approved):
lgtm :+1: 

Don't forget to squash

mfussenegger commented on extensions/functions/src/test/java/io/crate/operation/aggregation/HyperLogLogDistinctAggregationTest.java:
```diff
@@ -154,15 +155,20 @@ public void test_random_type_random_values() throws Exception {
         var type = RandomPicks.randomFrom(RandomizedContext.current().getRandom(), validTypes);
 
-        var data = TestingHelpers.getRandomsOfType(10000, 10000, type)
+        // For IP type we must ensure that provided values are handled similar to IpColumnReference.
+        Function ipFormatter = null;
+        if(type.id() == DataTypes.IP.id()) {
+            ipFormatter = o -> NetworkAddress.format(InetAddresses.forString((String) o));
+        }
+
+        var data = TestingHelpers.getRandomsOfType(10000, 10000, type, ipFormatter)
```
Instead of adding a workaround here we could adjust the data generation function to only return formatted values that match the `IpColumnReference` output: 

https://github.com/crate/crate/blob/6c57ce984602071fdadc16233732625ce5f0e9ec/server/src/testFixtures/java/io/crate/testing/DataTypeTesting.java#L93-L100

BaurzhanSakhariev replied:
Done. I was thinking about doing so at the beginning but had concern that some other tests using that generator might not expected specially formatted ip - but as long as tests are passing it's not a problem.

mfussenegger commented on server/src/testFixtures/java/io/crate/testing/TestingHelpers.java:
```diff
@@ -449,7 +449,11 @@ public static void assertCrateVersion(Object object, Version versionCreated, Ver
+                    values.add(dataType.sanitizeValue(valueHandler[0].apply(generator.get())));
+                } else {
+                    values.add(dataType.sanitizeValue(generator.get()));
+                }
```
See the other comment. I think we can avoid this workaround.
